expr/agg.go

```diff
@@ -0,0 +1,63 @@
+package expr
+
+import (
+	"errors"
+
+	"github.com/brimsec/zq/expr/agg"
+	"github.com/brimsec/zq/zng"
+)
+
+var (
+	ErrBadValue      = errors.New("bad value")
+	ErrFieldRequired = errors.New("field parameter required")
+)
+
+type Aggregator struct {
+	pattern agg.Pattern
+	expr    Evaluator
+	where   Evaluator
+}
+
+func NewAggregator(op string, expr, where Evaluator) (*Aggregator, error) {
+	pattern, err := agg.NewPattern(op)
+	if err != nil {
+		return nil, err
+	}
+	if expr == nil {
+		// Count is the only that has no argument so we just return
+		// true so it counts each value encountered.
+		expr = &Literal{zng.True}
+	}
+	return &Aggregator{
+		pattern: pattern,
+		expr:    expr,
+		where:   where,
+	}, nil
+}
+
+func (a *Aggregator) NewFunction() agg.Function {
+	return a.pattern()
+}
+
+func (a *Aggregator) Apply(f agg.Function, rec *zng.Record) {
+	if a.filter(rec) {
+		return
+	}
+	zv, err := a.expr.Eval(rec)
+	if err != nil {
+		//XXX
```

Comment:
Can this //XXX  be addressed before merging?

Reply:
Yeah good idea.  Bubbling up aggregator errors to group-by has been put off forever.  This refactor makes it easy now.  I'll fix it.